File: sms/src/main/java/com/mcw/sms/model/Shipment.java
package com.mcw.sms.model;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Shipment {

    private final String shipmentId;
    private State currentState;
    private final List<State> previousStates;

    public Shipment(String shipmentId, State currentState) {
        this.shipmentId = shipmentId;
        this.currentState = currentState;
        this.previousStates = new ArrayList<>();
    }

    public String getShipmentId() {
        return shipmentId;
    }

    public State getCurrentState() {
        return currentState;
    }

    public List<State> getPreviousStates() {
        return previousStates;
    }

    public void addNextState(State nextState) {
        previousStates.add(currentState);
        this.currentState = nextState;
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Shipment shipment)) return false;
        return Objects.equals(getShipmentId(), shipment.getShipmentId());
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(getShipmentId());
    }

    @Override
    public String toString() {
        return "Shipment{" +
                "shipmentId='" + shipmentId + '\'' +
                ", currentState=" + currentState +
                ", previousStates=" + previousStates +
                '}';
    }
}

File: sms_go/model/shipment.go
package model

type Shipment struct {
	id             string
	currentState   *State
	previousStates []*State
}

func (s *Shipment) AddNextState(state *State) {
	s.previousStates = append(s.previousStates, s.currentState)
	s.currentState = state
}

func (s *Shipment) GetCurrentState() *State {
	return s.currentState
}

func (s *Shipment) GetPreviousStates() []*State {
	return s.previousStates
}

func (s *Shipment) GetShipmentId() string {
	return s.id
}

func NewShipment(id string, currentState *State) *Shipment {
	return &Shipment{
		id:           id,
		currentState: currentState,
	}
}
